### PR TITLE
Optionally split regression tests into test groups

### PR DESCRIPTION
**Pull request by 4tXJ7f**
To prevent timing out on Travis, this commit adds the option to split
the regression tests into groups and run each group in a separate job.
The group of a test is determined by computing a checksum of its name.

### CODE REVIEW

File: test/regress/run_regression [hunk 1]
```diff
@@ -49,6 +49,16 @@ cvc4=$1
 benchmark_orig=$2
 benchmark="$benchmark_orig"
 
+if [ ! -z "$TEST_GROUP" ]; then
```
Comment (timothy-king) [inline test/regress/run_regression]: If we go with this solution, we should document the effect these have at the top of the file.

**Comment by 4tXJ7f** [comment]
I definitely hear your concerns about giving up control. Also, the proposed solution is not a very elegant one. To an extent, however, the regressX folders and `TEST_GROUP`s are orthogonal. The idea of this commit was to make it possible to split up one job into an arbitrary number of jobs (the slowest job is run as two jobs with the changes in `.travis.yml`). As far as I know, we are not using regressX for this purpose (if I am not mistaken, we only run regress0 on Travis). Either way, we should probably come up with a clear purpose for each of the regressX folders and a protocol for adding new tests, keeping in mind our limited resources.

**Comment by 4tXJ7f** [comment]
@mpreiner counted the test cases in each regressX directory:
```
regress0: 1538
regress1: 30
regress2: 36
regress3: 12
regress4: 6
```
(just copying it here because it might be relevant to the discussion)

**Comment by timothy-king** [comment]
> As far as I know, we are not using regressX for this purpose (if I am not mistaken, we only run regress0 on Travis).
Historical context: Originally the regressX directories were a scheme for running more expensive tests less frequently but still having them checked in and available. So regressX should be faster and run more often than regressX+1. But along the way regress0 became synonym with make check, which itself became the minimum for checking in code. Over time, all of the tests were just added to regress0. The incentives encourage this as adding a test makes a no big difference in personal running time and it is the natural home for always having the test run, which at the moment you definitely want. Tests that take so long can then just not listed in the Makefile.am files. So the original idea got out of sync with the maintenance of the tests.

So nothing really stops us from going back and cleaning these up and having cleaner regress groups with regress0 being tied to make check. We can also additionally have optional sharding. This is largely about preferences and making a benefits vs. time call. I suggest we talk about this at next week's meeting as it is a good candidate for folks expressing preferences.

**Comment by kbansal** [comment]
Uncomment ulimit line in this to detect long tests:

https://github.com/CVC4/CVC4/blob/master/test/regress/run_regression

I think the requirement was 1 sec in debug mode (on Dejan's machine? ) for it to be in regress0, but with increasing number of tests maybe that is not sufficient :)

File: test/regress/run_regression (hunk 1 above)
Comment (mpreiner) [inline test/regress/run_regression]: Instead of 'if [ ! -z  "$TEST_GROUP" ]' you could also use 'if [ -n "$TEST_GROUP" ]'.

Comment (4tXJ7f) [inline test/regress/run_regression]: Great point, my bash-fu failed me :)

**Pull request merged**